# import packages

In [1]:
import json
import re
import numpy as np
import pandas as pd
from write_load_data import load_data
from write_load_data import write_to_json

In [2]:
# load 'movie_dictionaries.json' and store as a variable named df. 
# If 'movie_dictionaries.json' does not exist in the directory this function will fail.
df = load_data('movie_dictionaries.json')

In [3]:
# [x for x in df if x['title']=='sacred planet']

In [4]:
# check the length of the dataframe loaded.
len(df)

447

In [5]:
#[clean] Testing to see if we are getting 'clean' versions of dictionaries.
# NOTE: To get the data format more standardized, all string fields are lower case.
[(index, x) for index,x in enumerate(df) if x['title']=="the vanishing prairie"]

[(24,
  {'title': 'the vanishing prairie',
   'Produced by': ['ben sharpsteen', 'walt disney'],
   'Written by': ['james algar', 'winston hibler'],
   'Narrated by': 'winston hibler',
   'Music by': 'paul j. smith',
   'Cinematography': 'n. paul kenworthy',
   'Edited by': 'lloyd l. richardson',
   'Production company': 'walt disney productions',
   'Distributed by': 'buena vista distribution',
   'Release date': ['august 17, 1954 ( 1954-08-17 )'],
   'Running time': ['60 minutes (vhs version)', '71 minutes (original)'],
   'Country': 'united states',
   'Language': 'english',
   'Box office': '$1.75 million (us and canadian rentals)'})]

In [6]:
# [clean] Testing to  see if we are getting 'clean' versions of dictionaries.
df[0:5]

[{'title': 'academy award review of walt disney cartoons',
  'Production company': 'walt disney productions',
  'Release date': ['may 19, 1937 ( 1937-05-19 )'],
  'Running time': '41 minutes (74 minutes 1966 release)',
  'Country': 'united states',
  'Language': 'english',
  'Box office': '$45.472'},
 {'title': 'snow white and the seven dwarfs',
  'Directed by': ['david hand (supervising)',
   'william cottrell',
   'wilfred jackson',
   'larry morey',
   'perce pearce',
   'ben sharpsteen'],
  'Produced by': 'walt disney',
  'Written by': ['ted sears',
   'richard creedon',
   'otto englander',
   'dick rickard',
   'earl hurd',
   'merrill de maris',
   'dorothy ann blank',
   'webb smith'],
  'Based on': ['snow white', 'brothers grimm'],
  'Starring': ['adriana caselotti',
   'lucille la verne',
   'harry stockwell',
   'roy atwell',
   'pinto colvig',
   'otis harlan',
   'scotty mattraw',
   'billy gilbert',
   'eddie collins',
   'moroni olsen',
   'stuart buchanan'],
  'Music by

# Clean the Data!

- ~~Standarize subscripts, Clean up references ~~
    - This was completed in 'task_2_scrape_movie_webpages.ipynb'. In the functions that created the movie dictionaries.
- Standarize Time 
    - Convert running time to an integer
    - convert data as date-time object
- Standarize Money
    - Convert budget
    - box office to numbers
- Standarize 'Starring' make a list

In [7]:
# Standardize Time
# Convert running time to an integer.
# Let's see how 'One Little Indian' turned out.
little_indian = [x for x in df if x['title']=="one little indian"]
little_indian

[{'title': 'one little indian',
  'Directed by': 'bernard mceveety',
  'Produced by': 'winston hibler',
  'Written by': 'harry spalding',
  'Starring': ['james garner',
   'vera miles',
   'pat hingle',
   'morgan woodward',
   'jodie foster'],
  'Music by': 'jerry goldsmith',
  'Cinematography': 'charles f. wheeler',
  'Edited by': 'robert stafford',
  'Production company': 'walt disney productions',
  'Distributed by': 'buena vista distribution',
  'Release date': ['june 20, 1973 ( 1973-06-20 )'],
  'Running time': '90 minutes',
  'Country': 'united states',
  'Language': 'english',
  'Box office': '$2 million'}]

In [8]:
# Navigate into the Running time value of 'One Little Indian'
running_time = little_indian[0]['Running time']
running_time

'90 minutes'

In [9]:
# How can we clean up the 'Running time' value to be an integer?
# Replace space (' ') and 'Minutes' with no values. 
# wrap the code with int() to change the type from string to an integer.
int(running_time.replace(' ', "").replace('minutes', ""))

90

In [10]:
# How many movies have the 'Running time' field
movies_with_runtime = [x for x in df if 'Running time' in x]
len(movies_with_runtime)

435

In [11]:
# another way to get the running time field from df. x.get(field_name)
# len([x.get('Running time') for x in df if x.get('Running time') is not None])

In [12]:
# strip running time if it is a string at '' 'min' (include the space so you don't have to strip it later). 
# movies_with_runtime = [{'title': x['title'], 'Running time': x['Running time'].split(' min', 1)[0]} if type(x['Running time']) is str else {'title':x['title']} for x in  movies_with_runtime ]
# movies_with_runtime

In [13]:
# remove the movies that don't have the runtime as a string:
# NOTE: '-' is a HYPHEN will take a copy of the output character to use in the split below.
# [{'title': x['title'], 'Running time': x['Running time'].split(' min', 1)[0]} for x in movies_with_runtime if x['title']=='zorro']

In [14]:
movies_with_runtime_int = [{'title': x['title'], 'Running time': int(x['Running time'].split(' min', 1)[0].split('–')[0])} for x in  movies_with_runtime if type(x['Running time']) is str]
movies_with_runtime_int

[{'title': 'academy award review of walt disney cartoons', 'Running time': 41},
 {'title': 'snow white and the seven dwarfs', 'Running time': 83},
 {'title': 'pinocchio', 'Running time': 88},
 {'title': 'fantasia', 'Running time': 126},
 {'title': 'the reluctant dragon', 'Running time': 74},
 {'title': 'dumbo', 'Running time': 64},
 {'title': 'bambi', 'Running time': 70},
 {'title': 'saludos amigos', 'Running time': 42},
 {'title': 'victory through air power', 'Running time': 65},
 {'title': 'the three caballeros', 'Running time': 71},
 {'title': 'make mine music', 'Running time': 75},
 {'title': 'song of the south', 'Running time': 94},
 {'title': 'fun and fancy free', 'Running time': 73},
 {'title': 'melody time', 'Running time': 75},
 {'title': 'so dear to my heart', 'Running time': 82},
 {'title': 'the adventures of ichabod andmr. toad', 'Running time': 68},
 {'title': 'cinderella', 'Running time': 74},
 {'title': 'treasure island', 'Running time': 96},
 {'title': 'alice in wonderl

In [15]:
# How many movies have a 'Running time' value that is an integer? 428/434
len(movies_with_runtime_int)

428

In [16]:
# Store the titles of movies with 'Running time' as an int into a list named title_runtime_int.
title_runtime_int = [x['title'] for x in movies_with_runtime_int]
title_runtime_int

['academy award review of walt disney cartoons',
 'snow white and the seven dwarfs',
 'pinocchio',
 'fantasia',
 'the reluctant dragon',
 'dumbo',
 'bambi',
 'saludos amigos',
 'victory through air power',
 'the three caballeros',
 'make mine music',
 'song of the south',
 'fun and fancy free',
 'melody time',
 'so dear to my heart',
 'the adventures of ichabod andmr. toad',
 'cinderella',
 'treasure island',
 'alice in wonderland',
 'the story of robin hood andhis merrie men',
 'peter pan',
 'the sword and the rose',
 'the living desert',
 'rob roy, the highland rogue',
 '20,000 leagues under the sea',
 'davy crockett, king of the wild frontier',
 'lady and the tramp',
 'the african lion',
 'the littlest outlaw',
 'the great locomotive chase',
 'davy crockett and the river pirates',
 'secrets of life',
 'westward ho the wagons!',
 'johnny tremain',
 'perri',
 'old yeller',
 'the light in the forest',
 'white wilderness',
 'tonka',
 'sleeping beauty',
 'the shaggy dog',
 "darby o'gill 

In [17]:
# The length of title_runtime_int equals movies_with_runtime_int
len(title_runtime_int)

428

In [18]:
# How many unique titles are there?
len({x:title_runtime_int.count(x) for x in title_runtime_int})

416

The count of unique movies that have an integer running time is 416.

In [19]:
# Rename the title_runtime_int list to only store the unique titles in title_runtime_int
title_runtime_int = list({x:title_runtime_int.count(x) for x in title_runtime_int}.keys())
len(title_runtime_int)

416

## Function: running_time(dictionary)

In [20]:
# Create a function that will take in a dictionary that contains the key 'Running time'.
# If the value type of 'Running time' is a str then change the value to an integer variable 'time'.
# return the 'time' variable
def running_time(dictionary):
    if type(dictionary['Running time']) is str:
        time = int(dictionary['Running time'].split(' ', 1)[0].split('–')[0])
       
    
    elif (type(dictionary['Running time']) is list) and (sum([x.find('minutes') for x in dictionary['Running time']])>0):
        time = [int(x.split(' ', 1)[0].split('–')[0]) for x in dictionary['Running time']][0]
     
    else:
        time =  dictionary['Running time']
            
    return time

In [21]:
# Testing running_time(dictionary) with 'vanishing praire'
test = df[[index for index, x in enumerate(df) if x['title'] == 'the vanishing prairie'][0]]
test

{'title': 'the vanishing prairie',
 'Produced by': ['ben sharpsteen', 'walt disney'],
 'Written by': ['james algar', 'winston hibler'],
 'Narrated by': 'winston hibler',
 'Music by': 'paul j. smith',
 'Cinematography': 'n. paul kenworthy',
 'Edited by': 'lloyd l. richardson',
 'Production company': 'walt disney productions',
 'Distributed by': 'buena vista distribution',
 'Release date': ['august 17, 1954 ( 1954-08-17 )'],
 'Running time': ['60 minutes (vhs version)', '71 minutes (original)'],
 'Country': 'united states',
 'Language': 'english',
 'Box office': '$1.75 million (us and canadian rentals)'}

In [22]:
[int(x.split(' min', 1)[0].split('–')[0]) for x in test['Running time']][0]

60

In [23]:
# Testing running_time(dictionary) function.
# Grab the dictionary for 'Savage Sam' and store the dictionary as variable, 'test'.
test = df[[index for index, x in enumerate(df) if x['title']=='savage sam'][0]]
test

{'title': 'savage sam',
 'Directed by': 'norman tokar',
 'Produced by': ['ron miller', 'walt disney'],
 'Written by': ['fred gipson', 'william turberg'],
 'Based on': 'savage sam by fred gipson',
 'Starring': ['brian keith', 'tommy kirk', 'kevin corcoran'],
 'Music by': 'oliver wallace',
 'Cinematography': 'edward colman',
 'Edited by': 'grant k. smith',
 'Production company': 'walt disney productions',
 'Distributed by': 'buena vista distribution',
 'Release date': 'june 1, 1963',
 'Running time': '103 minutes',
 'Country': 'united states',
 'Language': 'english',
 'Box office': '$3,000,000 (us/ canada)'}

In [24]:
# Testing running_time(dictionary) function with 'test' variable.
running_time(test)

103

In [25]:
# How many dictionaries in our list contain a title that is in the list title_runtime_int?
len([x for x in df if x['title'] in title_runtime_int ])

429

Question the data - Are there repeating dictionary titles maybe for remakes of movies?

In [26]:
# update df to change the keys of dictionaries that contain 'Running time' to integers using running_time(dictionary)
# Else - if there is no 'Running time' in the dictionary return the original dictionary

df = [{**x, 'Running time': running_time(x)} if 'Running time' in x else x for x in df ]

In [27]:
# How many movies do not have running time as an integer?
len([x for x in df if 'Running time' in x and type(x['Running time']) is not int])

4

In [28]:
# Which movies do not have 'Running time' as an integer?
[{'title': x['title'], 'Running time': x['Running time']} for x in df if 'Running time' in x and type(x['Running time']) is not int]

[{'title': 'the happiest millionaire',
  'Running time': ['los angeles', 'new york city', "director's cut"]},
 {'title': 'high school musical 3:  senior year', 'Running time': ['[1]']},
 {'title': 'jonas brothers:the 3d concert experience',
  'Running time': ['[1]', '[2]']},
 {'title': 'earth', 'Running time': ['[1]']}]

In [29]:
# Change running time of 'The Happiest Millionarie'
df[[index for index, x in enumerate(df) if x['title']=='the happiest millionaire'][0]]['Running time'] = 118

In [30]:
# Change running time of 'High School Musical 3:  Senior Year'
df[[index for index, x in enumerate(df) if x['title']=='high school musical 3:  senior year'][0]]['Running time'] = 111

In [31]:
# Change running time of 'Jonas Brothers: the 3D Concert Experience'
df[[index for index, x in enumerate(df) if x['title']=='jonas brothers:the 3d concert experience'][0]]['Running time'] = 76

In [32]:
# Change running time of 'Earth'
df[[index for index, x in enumerate(df) if x['title']=='earth'][0]]['Running time'] = 90

In [33]:
# Check that running time updated
[{'title': x['title'], 'Running time': x['Running time']} for x in df if x['title'] in ['earth', 'jonas brothers:the 3d concert experience', 'high school musical 3:  senior year', 'the happiest millionaire'] ]

[{'title': 'the happiest millionaire', 'Running time': 118},
 {'title': 'high school musical 3:  senior year', 'Running time': 111},
 {'title': 'jonas brothers:the 3d concert experience', 'Running time': 76},
 {'title': 'earth', 'Running time': 90}]

In [34]:
# Curious - What is the average 'Running time' value
sum([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])/len([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])

97.5816091954023

In [35]:
# What is the max running time?
max_time = max([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])
max_time

306

???? 306 minutes seems too long for a movie. Let's find out where this may be later.

In [36]:
# What is the minimum running time?
min_time = min(x['Running time'] for x in df if 'Running time' in x)
min_time

22

This seems too short for a feature length film.
\
\
Let's look at where these minimum and maximums are.

In [37]:
# find dictionaries where running time is equal to max_time. Store the index, and title in a list.
[(index, x['title']) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==max_time) ]

[(446, 'night at the museum')]

Going back to the wiki page for 'Night at the Museum' This is a movie that is set for future production. This movie is part of a series, and looks like the wiki pages shows running time as running time for all the movies. We can remove this dictionary from the df list.

In [38]:
# find dictionaries where running time is equal to min_time. Store the index, and title in a list.
[(index, x['title']) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==min_time)]

[(43, 'zorro'), (48, 'zorro')]

Going back to the wiki page for 'Zorro' these are both a part of series that aired on ABC. These are only episodes that are 22 minutes long. We can remove these dictionaries from the df list.


In [39]:
# Let's make sure there is only one 'Night at the Museum' movie before we remove the dictionary.
[(index, x['title']) for index, x in enumerate(df) if x['title']=='night at the museum']

[(446, 'night at the museum')]

In [40]:
# Let's make sure there are only two 'Zorro' movies before we remove the dictionaries.
[(index,x['title']) for index, x in enumerate(df) if x['title']=='zorro']

[(43, 'zorro'), (48, 'zorro')]

In [41]:
# Let's remove these movies from the list.
df = [x for x in df if x['title'] not in ['night at the museum', 'zorro']]

# the length should be 444 if we are dropping 3 movies.    
len(df)

444

In [42]:
# Now what is the movie with the max time?
max_time = max([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])
max_time

167

In [43]:
[(index, x) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==max_time) ]

[(302,
  {'title': "pirates of the caribbean:  at world's end",
   'Directed by': 'gore verbinski',
   'Produced by': 'jerry bruckheimer',
   'Written by': ['ted elliott', 'terry rossio'],
   'Based on': ['characters by ted elliott terry rossio stuart beattie jay wolpert',
    'ted elliott',
    'terry rossio',
    'stuart beattie',
    'jay wolpert',
    'pirates of the caribbean by walt disney'],
   'Starring': ['johnny depp',
    'orlando bloom',
    'keira knightley',
    'stellan skarsgård',
    'bill nighy',
    'chow yun-fat',
    'geoffrey rush',
    'jack davenport',
    'kevin r. mcnally',
    'jonathan pryce'],
   'Music by': 'hans zimmer',
   'Cinematography': 'dariusz wolski',
   'Edited by': ['craig wood', 'stephen rivkin'],
   'Production companies': ['walt disney pictures', 'jerry bruckheimer films'],
   'Distributed by': 'buena vista pictures',
   'Release date': ['may 19, 2007 ( 2007-05-19 ) ( disneyland resort )',
    'may 25, 2007 ( 2007-05-25 ) (united states)'],
 

In [44]:
# Now what is the movie with the min time?
min_time = min(x['Running time'] for x in df if 'Running time' in x)
min_time

40

In [45]:
[(index, x) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==min_time)]

[(273,
  {'title': 'sacred planet',
   'Directed by': ['jon long', 'hairul salleh askor'],
   'Produced by': ['jon long', 'karen fernandez long'],
   'Written by': ['karen fernandez long', 'jon long'],
   'Starring': ['arapata mckay',
    'tsaan ciqae',
    'mae tui',
    'cy peck jr.',
    'mutang urud'],
   'Narrated by': 'robert redford',
   'Cinematography': 'william reeve',
   'Edited by': 'jon long',
   'Production company': 'walt disney pictures',
   'Distributed by': 'buena vista pictures',
   'Release date': ['april 22, 2004 ( 2004-04-22 )'],
   'Running time': 40,
   'Countries': ['canada', 'malaysia', 'united states'],
   'Language': 'english',
   'Box office': '$1,108,356'}),
 (291,
  {'title': 'roving mars',
   'Directed by': 'george butler',
   'Produced by': ['frank marshall'],
   'Written by': ['robert andrus', 'george butler'],
   'Narrated by': 'paul newman (introduction only)',
   'Music by': ['philip glass', 'sigur rós'],
   'Cinematography': 't.c. christensen',
   

In [46]:
# We now have the Running time field cleaned up.
# let's save this to a 'cleaned' version - movie_dictionaries_cleaned.json

write_to_json(df,'movie_dictionaries_cleaned.json')

In [47]:
df=load_data('movie_dictionaries_cleaned.json')
len(df)

444

In [48]:
[(index, x['title']) for index,x in enumerate(df) if x.get('Running time', 'N\A')=='N\A']

[(324, 'lilly the witch:the dragon and the magic book'),
 (327, 'walt & el grupo'),
 (344, 'anaganaga o dheerudu'),
 (401, 'ghost of the mountains'),
 (436, 'cruella'),
 (437, 'jungle cruise'),
 (438, 'the beatles: get back'),
 (439, 'encanto'),
 (440, 'the little mermaid'),
 (441, 'peter pan & wendy'),
 (442, 'home alone'),
 (443, 'shrunk')]

Looking at the wiki pages for these movies the running time is either not listed or the movie is 'in production' phase

# Cleaning task - Standardize budget and box office fields to be float data type.

In [49]:
# movie has 'Budget' data
len([x.get('Budget') for x in df if x.get('Budget') is not None])

284

In [50]:
# movie has 'Box office' data
len([x.get('Box office') for x in df if x.get('Box office') is not None])

368

In [51]:
# movie has both 'Box office' and 'Budget data'
len([(x.get('Budget'), x.get('Box office')) for x in df if x.get('Budget') is not None and x.get('Box office') is not None])

272

In [52]:
# what does the format of the budget key look like?
[(i,x.get('Budget')) for i,x in enumerate(df) if x.get('Budget') is not None]

[(1, '$1.49 million'),
 (2, '$2.6 million'),
 (3, '$2.28 million'),
 (4, '$600,000'),
 (5, '$950,000'),
 (6, '$858,000'),
 (8, '$788,000'),
 (10, '$1.35 million'),
 (11, '$2.125 million'),
 (13, '$1.5 million'),
 (14, '$1.5 million'),
 (16, '$2.9 million'),
 (17, '$1,800,000'),
 (18, '$3 million'),
 (20, '$4 million'),
 (21, '$2 million'),
 (22, '$300,000'),
 (23, '$1.8 million'),
 (25, '$5 million'),
 (27, '$4 million'),
 (34, '$700,000'),
 (40, '$6 million'),
 (41, 'under $1 million or $1,250,000'),
 (43, '$2 million'),
 (46, '$2.5 million'),
 (49, '$4 million'),
 (50, '$3.6 million'),
 (55, '$3 million'),
 (57, '$3 million'),
 (67, '$3 million'),
 (72, '$4.4–6 million'),
 (84, '$4 million'),
 (86, '$5 million'),
 (91, '$5 million'),
 (98, '$4 million'),
 (102, '$6.3 million'),
 (111, '$5 million'),
 (116, '$8 million'),
 (122, 'au$1 million'),
 (127, '$5 million'),
 (131, '$7.5 million'),
 (133, '$10 million'),
 (136, '$3.5 to 4 million'),
 (140, '$5.25 million'),
 (141, '$20 millio

- Let's strip out `$` - strip('$')
- Let's strip out 'million' and the space before it so we don't have to strip the space later - .strip(' million')
- For the movies that are in hundred thousands replace ',' with '' - .replace(',',"")

In [53]:
budget_int_list = []
could_not_convert = []
for i,x in enumerate(df):
    if x.get('Budget') is not None and type(x.get('Budget')) is str:
        try:
            budget = float(x.get('Budget').strip('$').split(' million', 1)[0].replace(',',"").split('–',1)[0].split('-')[0].split('to',1)[0])
            
            
            if budget < 500.0:
                budget =budget * (1000000)
            
            budget_int_list.append(budget)
            df[i]['Budget']=budget
            print(i,budget)
        
        except Exception as e:
            could_not_convert.append({i:{'title':x.get('title'),'Budget':x.get('Budget')}})
            print(e, i, x.get('title'), x.get('Budget'))
            continue
            
    elif x.get('Budget') is not None and type(x.get('Budget')) is list:
        try:
            budget=float(x.get('Budget')[0].strip('$').split(' million', 1)[0].replace(',',"").split('–',1)[0].split('-')[0].split('to',1)[0])
            
            if budget < 500.0:
                budget =budget * (1000000)
            
            budget_int_list.append(budget)
            df[i]['Budget']=budget
            print(i,budget)
        
        except Exception as e:
            could_not_convert.append({i:{'title':x.get('title'),'Budget':x.get('Budget')}})
            print(e, i, x.get('title'), x.get('Budget'))
            continue

1 1490000.0
2 2600000.0
3 2280000.0
4 600000.0
5 950000.0
6 858000.0
8 788000.0
10 1350000.0
11 2125000.0
13 1500000.0
14 1500000.0
16 2900000.0
17 1800000.0
18 3000000.0
20 4000000.0
21 2000000.0
22 300000.0
23 1800000.0
25 5000000.0
27 4000000.0
34 700000.0
40 6000000.0
could not convert string to float: 'under $1' 41 the shaggy dog under $1 million or $1,250,000
43 2000000.0
46 2500000.0
49 4000000.0
50 3600000.0
55 3000000.0
57 3000000.0
67 3000000.0
72 4400000.0
84 4000000.0
86 5000000.0
91 5000000.0
98 4000000.0
102 6300000.0
111 5000000.0
116 8000000.0
could not convert string to float: 'au$1' 122 ride a wild pony au$1 million
127 5000000.0
131 7500000.0
133 10000000.0
136 3500000.0
140 5250000.0
141 20000000.0
144 9000000.0
146 6000000.0
147 20000000.0
150 18000000.0
151 12000000.0
152 14000000.0
could not convert string to float: 'us$ 17' 154 tron us$ 17 million
155 5000000.0
could not convert string to float: 'unknown' 156 trenchcoat unknown
157 20000000.0
158 11000000.0
159 

In [54]:
budget_int_list.sort()
budget_int_list

[300000.0,
 600000.0,
 700000.0,
 788000.0,
 858000.0,
 950000.0,
 1000000.0,
 1350000.0,
 1490000.0,
 1500000.0,
 1500000.0,
 1800000.0,
 1800000.0,
 2000000.0,
 2000000.0,
 2125000.0,
 2280000.0,
 2500000.0,
 2600000.0,
 2900000.0,
 3000000.0,
 3000000.0,
 3000000.0,
 3000000.0,
 3000000.0,
 3500000.0,
 3600000.0,
 4000000.0,
 4000000.0,
 4000000.0,
 4000000.0,
 4000000.0,
 4400000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5000000.0,
 5250000.0,
 6000000.0,
 6000000.0,
 6300000.0,
 6500000.0,
 7000000.0,
 7500000.0,
 8000000.0,
 8000000.0,
 8000000.0,
 8000000.0,
 9000000.0,
 9000000.0,
 10000000.0,
 10000000.0,
 10000000.0,
 11000000.0,
 11000000.0,
 12000000.0,
 12000000.0,
 12000000.0,
 12000000.0,
 12500000.0,
 13000000.0,
 13000000.0,
 14000000.0,
 14000000.0,
 14000000.0,
 14000000.0,
 15000000.0,
 15000000.0,
 15000000.0,
 15000000.0,
 15000000.0,
 15000000.0,
 15000000.0,

In [55]:
len(could_not_convert)

13

In [56]:
could_not_convert

[{41: {'title': 'the shaggy dog', 'Budget': 'under $1 million or $1,250,000'}},
 {122: {'title': 'ride a wild pony', 'Budget': 'au$1 million'}},
 {154: {'title': 'tron', 'Budget': 'us$ 17 million'}},
 {156: {'title': 'trenchcoat', 'Budget': 'unknown'}},
 {166: {'title': 'the man from snowy river ii', 'Budget': 'a$8.7 million'}},
 {254: {'title': 'spirited away', 'Budget': ['¥', '[2]', '[3]', '[4]']}},
 {284: {'title': "howl's moving castle", 'Budget': ['¥']}},
 {326: {'title': 'ponyo', 'Budget': ['¥', 'us$']}},
 {344: {'title': 'anaganaga o dheerudu',
   'Budget': '₹ 27 crore (us$3.8 million)'}},
 {374: {'title': 'khoobsurat', 'Budget': '₹ 230 million (us$3.2 million)'}},
 {383: {'title': 'abcd 2', 'Budget': '₹520 million'}},
 {395: {'title': 'dangal', 'Budget': '₹ 70 crore'}},
 {402: {'title': 'jagga jasoos', 'Budget': ['[2]']}}]

In [57]:
[{i:{x.get('title'): x.get('Budget')}} for i,x in enumerate(df) if type(x.get('Budget')) is not float and x.get('Budget') is not None]

[{41: {'the shaggy dog': 'under $1 million or $1,250,000'}},
 {122: {'ride a wild pony': 'au$1 million'}},
 {154: {'tron': 'us$ 17 million'}},
 {156: {'trenchcoat': 'unknown'}},
 {166: {'the man from snowy river ii': 'a$8.7 million'}},
 {254: {'spirited away': ['¥', '[2]', '[3]', '[4]']}},
 {284: {"howl's moving castle": ['¥']}},
 {326: {'ponyo': ['¥', 'us$']}},
 {344: {'anaganaga o dheerudu': '₹ 27 crore (us$3.8 million)'}},
 {374: {'khoobsurat': '₹ 230 million (us$3.2 million)'}},
 {383: {'abcd 2': '₹520 million'}},
 {395: {'dangal': '₹ 70 crore'}},
 {402: {'jagga jasoos': ['[2]']}}]

We will go ahead manually manipulate these budgets based on their wiki pages. If there is an austrailian (au) currency we will default to us currency.

In [58]:
df[41]['Budget'] = 1250000.0
df[122]['Budget'] = 1000000.0
df[154]['Budget'] = 17000000.0
df[156]['Budget'] = 'NA'
df[166]['Budget'] = 8700000.0
df[254]['Budget'] = 15000000.0
df[284]['Budget'] = 24000000.0
df[326]['Budget'] = 34000000.0
df[344]['Budget'] = 3800000.0
df[374]['Budget'] = 3200000.0
df[383]['Budget'] = 'NA'
df[395]['Budget'] = 'NA'
df[402]['Budget'] = 'NA'

In [59]:
# how many movies have a budget value that is now a float?
len([x for x in df if type(x.get('Budget')) is float])

280

This makes sense because we manipulated 4 movies to have a value of 'NA' that did not have a budget value of float.

In [60]:
# How many movies have a Budget key?
len([x for x in df if x.get('Budget') is not None])

284

This makes sense because this is the total amount of movies that have a budget type of float, and includes the movies that have Budget as 'NA' that we manipulated.

In [61]:
# How many movies DO NOT have a budget key?
len([x for x in df if x.get('Budget') is None])

160

There are 284 movies that have a budget key and 160 that do not. This totals 444 movies that is the total amount in df. :)

Now let's look at cleanning up the box office field.

In [62]:
# How is the box office field formatted?
[x['Box office'] for x in df if x.get('Box office') is not None]

['$45.472',
 '$418 million',
 '$164 million',
 '$76.4–$83.3 million',
 '$960,000 (worldwide rentals)',
 '$1.3 million (est. united states/canada rentals, 1941)',
 '$267.4 million',
 '$1,135,000 (worldwide rentals)',
 '$799,000',
 '$3,355,000 (worldwide rentals)',
 '$3.275 million (worldwide rentals)',
 '$65 million',
 '$3,165,000 (worldwide rentals)',
 '$2,560,000 (worldwide rentals)',
 '$3.7 million (u.s. rental) $575,000 (foreign rental)',
 '$1,625,000 (worldwide rentals)',
 '$263.6 million',
 '$4,100,000 (worldwide rentals)',
 '$5.6 million (us, 1951)',
 '$2.1 million (us rentals)',
 '$87.4 million',
 '$1 million (us)',
 '$2.6 million (us)',
 '$1.75 million (us and canadian rentals)',
 '$28.2 million',
 '$2,150,000 (us)',
 '$187 million',
 '$2.1 million (us)',
 '$1.6 million (us)',
 '$1.7 million (us)',
 '$2.75 million (us)',
 '$1.75 million (us rentals)',
 '$6,250,000 (us/canada rentals)',
 '$1.8 million (est. us/ canada rentals)',
 '$2.5 million (est. us/ canada rentals)',
 '$51.6

In [63]:
# # Looks like we can work this similar to before.

# boxoffice_int_list = []
# could_not_convert = []
# for i,x in enumerate(df):
#     if x.get('Box office') is not None and type(x.get('Box office')) is str:
#         try:
#             boxoffice = float(x.get('Box office').split(' million', 1)[0].split('(worldwide rentals)',1)[0].strip('$').replace(',',"").split('–$',1)[0])
            
            
#             if boxoffice < 1000.0:
#                 boxoffice =boxoffice * (1000000)
            
#             boxoffice_int_list.append((boxoffice,x.get('Box office')))
# #           df[i]['Box office']=boxoffice
#             print(i,boxoffice)
        
#         except Exception as e:
#             could_not_convert.append({i:{'title':x.get('title'),'Box office':x.get('Box office')}})
#             print(e, i, x.get('title'), x.get('Box office'))
#             continue
    
            
#     elif x.get('Box office') is not None and type(x.get('Box office')) is list:
#         try:
#             boxoffice=float(x.get('Box office')[0].split(' million',1)[0].split('(worldwide rentals)',1)[0].strip('$').replace(',',"").split('–$',1)[0])
            
#             if boxoffice < 1000.0:
#                 boxoffice = boxoffice * (1000000)
            
#             boxoffice_int_list.append((boxoffice,x.get('Box office')))
# #             df[i]['Box office']=boxoffice
#             print(i,budget)
        
#         except Exception as e:
#             could_not_convert.append({i:{'title':x.get('title'),'Box office':x.get('Box office')}})
#             print(e, i, x.get('title'), x.get('Box office'))
#             continue


In [64]:
# boxoffice_int_list.sort()
# boxoffice_int_list

#### Try using regex compiler

In [65]:
box_office_values = [x.get('Box office') for i,x in enumerate(df) if type(x.get('Box office')) is str]
box_office_values

['$45.472',
 '$418 million',
 '$164 million',
 '$76.4–$83.3 million',
 '$960,000 (worldwide rentals)',
 '$1.3 million (est. united states/canada rentals, 1941)',
 '$267.4 million',
 '$1,135,000 (worldwide rentals)',
 '$799,000',
 '$3,355,000 (worldwide rentals)',
 '$3.275 million (worldwide rentals)',
 '$65 million',
 '$3,165,000 (worldwide rentals)',
 '$2,560,000 (worldwide rentals)',
 '$3.7 million (u.s. rental) $575,000 (foreign rental)',
 '$1,625,000 (worldwide rentals)',
 '$263.6 million',
 '$4,100,000 (worldwide rentals)',
 '$5.6 million (us, 1951)',
 '$2.1 million (us rentals)',
 '$87.4 million',
 '$1 million (us)',
 '$2.6 million (us)',
 '$1.75 million (us and canadian rentals)',
 '$28.2 million',
 '$2,150,000 (us)',
 '$187 million',
 '$2.1 million (us)',
 '$1.6 million (us)',
 '$1.7 million (us)',
 '$2.75 million (us)',
 '$1.75 million (us rentals)',
 '$6,250,000 (us/canada rentals)',
 '$1.8 million (est. us/ canada rentals)',
 '$2.5 million (est. us/ canada rentals)',
 '$51.6

In [66]:
amount_pattern=re.compile('.*?\$.*?(?P<amount>[0-9,.]+)')
amount_pattern

re.compile(r'.*?\$.*?(?P<amount>[0-9,.]+)', re.UNICODE)

In [67]:
place_holder_pattern=re.compile('.*?(?P<mb>million|billion)')
place_holder_pattern

re.compile(r'.*?(?P<mb>million|billion)', re.UNICODE)

In [68]:
bov=box_office_values[1]
bov

'$418 million'

In [69]:
place_holder_pattern.search(bov).groups()

('million',)

In [70]:
[(x,{'pcv': place_holder_pattern.search(x).groups()[0]}) for x in box_office_values if place_holder_pattern.search(x) is not None]

[('$418 million', {'pcv': 'million'}),
 ('$164 million', {'pcv': 'million'}),
 ('$76.4–$83.3 million', {'pcv': 'million'}),
 ('$1.3 million (est. united states/canada rentals, 1941)',
  {'pcv': 'million'}),
 ('$267.4 million', {'pcv': 'million'}),
 ('$3.275 million (worldwide rentals)', {'pcv': 'million'}),
 ('$65 million', {'pcv': 'million'}),
 ('$3.7 million (u.s. rental) $575,000 (foreign rental)', {'pcv': 'million'}),
 ('$263.6 million', {'pcv': 'million'}),
 ('$5.6 million (us, 1951)', {'pcv': 'million'}),
 ('$2.1 million (us rentals)', {'pcv': 'million'}),
 ('$87.4 million', {'pcv': 'million'}),
 ('$1 million (us)', {'pcv': 'million'}),
 ('$2.6 million (us)', {'pcv': 'million'}),
 ('$1.75 million (us and canadian rentals)', {'pcv': 'million'}),
 ('$28.2 million', {'pcv': 'million'}),
 ('$187 million', {'pcv': 'million'}),
 ('$2.1 million (us)', {'pcv': 'million'}),
 ('$1.6 million (us)', {'pcv': 'million'}),
 ('$1.7 million (us)', {'pcv': 'million'}),
 ('$2.75 million (us)', {'pc

In [71]:
amount_pattern.match(box_office_values[3]).groups()

('76.4',)

In [72]:
place_holder_pattern.match(box_office_values[1]).groups()

('million',)

In [73]:
[x if place_holder_pattern.search(x) is None else (i, place_holder_pattern.search(x).groupdict(),x) for i,x in enumerate(box_office_values)]

['$45.472',
 (1, {'mb': 'million'}, '$418 million'),
 (2, {'mb': 'million'}, '$164 million'),
 (3, {'mb': 'million'}, '$76.4–$83.3 million'),
 '$960,000 (worldwide rentals)',
 (5,
  {'mb': 'million'},
  '$1.3 million (est. united states/canada rentals, 1941)'),
 (6, {'mb': 'million'}, '$267.4 million'),
 '$1,135,000 (worldwide rentals)',
 '$799,000',
 '$3,355,000 (worldwide rentals)',
 (10, {'mb': 'million'}, '$3.275 million (worldwide rentals)'),
 (11, {'mb': 'million'}, '$65 million'),
 '$3,165,000 (worldwide rentals)',
 '$2,560,000 (worldwide rentals)',
 (14,
  {'mb': 'million'},
  '$3.7 million (u.s. rental) $575,000 (foreign rental)'),
 '$1,625,000 (worldwide rentals)',
 (16, {'mb': 'million'}, '$263.6 million'),
 '$4,100,000 (worldwide rentals)',
 (18, {'mb': 'million'}, '$5.6 million (us, 1951)'),
 (19, {'mb': 'million'}, '$2.1 million (us rentals)'),
 (20, {'mb': 'million'}, '$87.4 million'),
 (21, {'mb': 'million'}, '$1 million (us)'),
 (22, {'mb': 'million'}, '$2.6 million (u

In [74]:
# create a dictionary to store the regex compilers to search for amounts and million/billion.

bol = []

for i,x in enumerate(box_office_values):
    box_office_dict={}
    
    if amount_pattern.search(x) is not None:
        box_office_dict['amount'] = amount_pattern.search(x).groups()[0].strip('$')
    if place_holder_pattern.search(x) is not None:
        box_office_dict['place_holder'] = place_holder_pattern.search(x).groups()[0]
    
    if box_office_dict.get('amount') is not None:
        box_office_dict['value'] = float(box_office_dict['amount'].replace(",",""))
        
        if box_office_dict.get('place_holder') is not None:
            if box_office_dict.get('place_holder') == 'million':
                box_office_dict['value'] = box_office_dict['value']*1000000
                
            else:
                box_office_dict['value'] = box_office_dict['value']*1000000000
            
    bol.append(box_office_dict)
bol

[{'amount': '45.472', 'value': 45.472},
 {'amount': '418', 'place_holder': 'million', 'value': 418000000.0},
 {'amount': '164', 'place_holder': 'million', 'value': 164000000.0},
 {'amount': '76.4', 'place_holder': 'million', 'value': 76400000.0},
 {'amount': '960,000', 'value': 960000.0},
 {'amount': '1.3', 'place_holder': 'million', 'value': 1300000.0},
 {'amount': '267.4', 'place_holder': 'million', 'value': 267399999.99999997},
 {'amount': '1,135,000', 'value': 1135000.0},
 {'amount': '799,000', 'value': 799000.0},
 {'amount': '3,355,000', 'value': 3355000.0},
 {'amount': '3.275', 'place_holder': 'million', 'value': 3275000.0},
 {'amount': '65', 'place_holder': 'million', 'value': 65000000.0},
 {'amount': '3,165,000', 'value': 3165000.0},
 {'amount': '2,560,000', 'value': 2560000.0},
 {'amount': '3.7', 'place_holder': 'million', 'value': 3700000.0},
 {'amount': '1,625,000', 'value': 1625000.0},
 {'amount': '263.6', 'place_holder': 'million', 'value': 263600000.00000003},
 {'amount':

# Function: create_re_pattern()

In [75]:
def create_re_patterns():
    amount_pattern=re.compile('.*?\$.*?(?P<amount>[0-9,.]+)')
    place_holder_pattern = place_holder_pattern=re.compile('.*?(?P<mb>million|billion)')
    
    return (amount_pattern, place_holder_pattern)
    

# Function: create_bo_values(dictionary)

In [76]:
def create_bo_values(dictionary):
    
    amount_pattern, place_holder_pattern = create_re_patterns()
    
    box_office_dict = {}
    
    if dictionary.get('Box office') is not None:
        if type(dictionary.get('Box office')) is str:
            if amount_pattern.search(dictionary['Box office']) is not None:
                box_office_dict['amount'] = amount_pattern.search(dictionary['Box office']).groups()[0].strip('$')
            if place_holder_pattern.search(dictionary['Box office']) is not None:
                box_office_dict['place_holder'] = place_holder_pattern.search(dictionary['Box office']).groups()[0]
            
            if box_office_dict.get('amount') is not None:
                box_office_dict['value'] = float(box_office_dict['amount'].replace(",",""))
            
            
                if box_office_dict.get('place_holder') is not None:
                    if box_office_dict.get('place_holder') == 'million':
                        box_office_dict['value'] = box_office_dict['value']*1000000
                
                    else:
                        box_office_dict['value'] = box_office_dict['value']*1000000000
            
            if box_office_dict.get('value') is not None:
                dictionary['bov'] = box_office_dict['value']
            
    return dictionary

In [77]:
df[23]

{'title': 'rob roy, the highland rogue',
 'Directed by': 'harold french',
 'Produced by': ['perce pearce', 'walt disney'],
 'Written by': 'lawrence edward watkin',
 'Starring': ['richard todd',
  'glynis johns',
  'james robertson justice',
  'michael gough',
  'finlay currie',
  'geoffrey keen'],
 'Music by': 'cedric thorpe davie',
 'Cinematography': 'guy green',
 'Edited by': 'geoffrey foot',
 'Production company': 'walt disney productions',
 'Distributed by': 'rko radio pictures',
 'Release date': ['26 october 1953 ( 1953-10-26 ) (premiere- london )',
  '4 february 1954 ( 1954-02-04 ) ( us )'],
 'Running time': 81,
 'Countries': ['united kingdom', 'united states'],
 'Language': 'english',
 'Budget': 1800000.0}

In [78]:
# Create another df object to hold all dictionaries and the new 'bov' values
df_bov = [(create_bo_values(x)) for i,x in enumerate(df)]

In [79]:
len([x.get('bov') for i,x in enumerate(df_bov) if x.get('bov') is not None])

361

In [80]:
len([x.get('Box office') for x in df_bov if x.get('Box office') is not None])

368

In [81]:
len([x for x in df_bov if x.get('Box office') is not None and x.get('bov') is None])

7

In [82]:
[(i,x.get('title'),x.get('Box office')) for i,x in enumerate(df_bov) if x.get('Box office') is not None and x.get('bov') is None]

[(42, "darby o'gill and the little people", ['[2]', '[3]']),
 (113, 'herbie rides again', ['[1]']),
 (141, 'the black hole', ['[1]']),
 (284, "howl's moving castle", ['¥', '[3]']),
 (383, 'abcd 2', 'est. ₹1.65 billion'),
 (395, 'dangal', ['[4]', '[5]', '[6]', '[7]', '[8]']),
 (402, 'jagga jasoos', '₹ 833.5 million')]

 going to the wiki pages for the movies we will insert the values for the movies we are able to insert.

In [83]:
df_bov[42]['bov'] = 2600000.0
df_bov[113]['bov'] = 30800000.0
df_bov[141]['bov'] = 35800000.0
df_bov[284]['bov'] = 236000000.0
df_bov[395]['bov'] = 311000000.0
